### PR TITLE
[AIRFLOW-2948] Arg check & better doc - SSHOperator & SFTPOperator

### PR DESCRIPTION
**Pull request by XD-DENG**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-2948
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

There may be different combinations of arguments, and some processings are being done 'silently', while users may not be fully aware of them.

For example
- User only needs to provide either `ssh_hook` or `ssh_conn_id`, while this is not clear in doc
- if both provided, `ssh_conn_id` will be ignored.
- if `remote_host` is provided, it will replace the `remote_host` which wasndefined in `ssh_hook` or predefined in the connection of `ssh_conn_id`

These should be documented clearly to ensure it's transparent to the users.

log.info() should also be used to remind users and provide clear logs.

### Tests

- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [ ] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.

### Code Quality

- [x] Passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

File: airflow/contrib/operators/sftp_operator.py [hunk 1]
```diff
@@ -77,13 +81,19 @@ def __init__(self,
...
         try:
-            if self.ssh_conn_id and not self.ssh_hook:
-                self.ssh_hook = SSHHook(ssh_conn_id=self.ssh_conn_id)
+            if self.ssh_conn_id:
+                if self.ssh_hook:
+                    self.log.info("ssh_conn_id is ignored when ssh_hook is provided.")
+                else:
+                    self.ssh_hook = SSHHook(ssh_conn_id=self.ssh_conn_id)
 
             if not self.ssh_hook:
```
Comment (feng-tao) [inline airflow/contrib/operators/sftp_operator.py]: in this case do we still need this line?

File: airflow/contrib/operators/ssh_operator.py [hunk 2]
```diff
@@ -68,14 +72,20 @@ def __init__(self,
...
-                self.ssh_hook = SSHHook(ssh_conn_id=self.ssh_conn_id,
-                                        timeout=self.timeout)
+            if self.ssh_conn_id:
+                if self.ssh_hook:
+                    self.log.info("ssh_conn_id is ignored when ssh_hook is provided.")
+                else:
+                    self.ssh_hook = SSHHook(ssh_conn_id=self.ssh_conn_id,
+                                            timeout=self.timeout)
 
             if not self.ssh_hook:
```
Comment (feng-tao) [inline airflow/contrib/operators/ssh_operator.py]: same

File: airflow/contrib/operators/sftp_operator.py [hunk 3]
```diff
@@ -77,13 +81,19 @@ def __init__(self,
     def execute(self, context):
         file_msg = None
         try:
-            if self.ssh_conn_id and not self.ssh_hook:
-                self.ssh_hook = SSHHook(ssh_conn_id=self.ssh_conn_id)
+            if self.ssh_conn_id:
+                if self.ssh_hook:
+                    self.log.info("ssh_conn_id is ignored when ssh_hook is provided.")
+                else:
+                    self.ssh_hook = SSHHook(ssh_conn_id=self.ssh_conn_id)
```
Comment (feng-tao) [inline airflow/contrib/operators/sftp_operator.py]: why don't we set timeout for this hook but the one in SSHOperator?

File: airflow/contrib/operators/sftp_operator.py [hunk 4]
```diff
@@ -77,13 +81,19 @@ def __init__(self,
     def execute(self, context):
         file_msg = None
         try:
-            if self.ssh_conn_id and not self.ssh_hook:
-                self.ssh_hook = SSHHook(ssh_conn_id=self.ssh_conn_id)
+            if self.ssh_conn_id:
+                if self.ssh_hook:
```
Comment (feng-tao) [inline airflow/contrib/operators/sftp_operator.py]: And I wonder we should check the self.ssh_hook is an instance for SSHHook.

File: airflow/contrib/operators/sftp_operator.py (hunk 1 above)
Comment (XD-DENG) [reply airflow/contrib/operators/sftp_operator.py]: This line is still necessary: what if neither `ssh_conn_id` nor `ssh_hook` is provided?

File: airflow/contrib/operators/ssh_operator.py (hunk 2 above)
Comment (XD-DENG) [reply airflow/contrib/operators/ssh_operator.py]: The same: we need to guard the situation in which neither `ssh_conn_id` nor `ssh_hook` is provided.

File: airflow/contrib/operators/sftp_operator.py (hunk 4 above)
Comment (XD-DENG) [reply airflow/contrib/operators/sftp_operator.py]: Agree on this.

I will add the instance check and update tests for this (to test whether `ssh_hook` of wrong instance can be captured).

File: airflow/contrib/operators/sftp_operator.py (hunk 3 above)
Comment (XD-DENG) [reply airflow/contrib/operators/sftp_operator.py]: I have noticed this as well. I'm following the existing implementation.

`timeout` was added for `SSHOperator` in PR https://github.com/apache/incubator-airflow/pull/3666 . I'm not sure if there is any good reason for which `timeout` was added for `SSHOperator` but not added for the `SSHHook` in `SFTPOperator`. Giving I'm not confident about the reason behind, I don't intend to make change on this in this PR as well.

**Comment by XD-DENG** [comment]
Hi @feng-tao , I have
- Added instance checking for argument `ssh_hook`. If `ssh_hook` given by user is not of class "SSHHook", it would be handled properly.
- Updated the tests for both operators accordingly (covered different combinations of arguments `ssh_hook` and `ssh_conn_id`). Both can pass.

The CI is failing due to:
- test_poke (tests.contrib.sensors.test_aws_redshift_cluster_sensor.`TestAwsRedshiftClusterSensor`) ... ERROR
- test_poke_cluster_not_found (tests.contrib.sensors.test_aws_redshift_cluster_sensor.`TestAwsRedshiftClusterSensor`) ... ERROR
- test_poke_false (tests.contrib.sensors.test_aws_redshift_cluster_sensor.`TestAwsRedshiftClusterSensor`) ... ERROR

None of these test cases were touched in this PR. Seems it's exception like `Unable to locate credentials`. Not sure what is happening. Similar issue is happening to a few other PRs.

**Comment by XD-DENG** [comment]
CI succeeds now after Fokko fixs the boto3 dependency issue in CI tests.

**Comment by XD-DENG** [comment]
Hi @feng-tao may you take another look? Thanks!

**Comment by feng-tao** [comment]
lgtm @XD-DENG , thanks for the pr

**Pull request merged**